scheds/rust/scx_rustland/src/bpf/alloc.rs

```diff
@@ -44,6 +46,71 @@ static MEMORY: RustLandMemory = RustLandMemory {
     allocation_map: Mutex::new([false; NUM_BLOCKS]),
 };
 
+// State of special sysctl VM settings.
+//
+// This is used to save the previous state of some procfs settings that must be changed by the
+// user-space scheduler.
+struct VmSettings {
+    // We cannot allow page faults in the user-space scheduler, otherwise we may hit deadlock
+    // conditions: a kthread may need to run to resolve the page fault, but the user-space
+    // scheduler is waiting on the page fault to be resolved => deadlock.
+    //
+    // To prevent this from happening automatically enforce vm.compact_unevictable_allowed=0 when
+    // the scheduler is running, to disable compaction of unevictable memory pages and make sure
+    // that the scheduler never faults.
+    //
+    // The original value will be restored when the user-space scheduler exits.
+    compact_unevictable_allowed: i32,
+}
+
+impl VmSettings {
+    // Return the content of a procfs file as i32.
+    fn read_procfs(&self, file_path: &str) -> i32 {
+        let file = File::open(file_path).expect(&format!("Failed to open {}", file_path));
+        let reader = BufReader::new(file);
+
+        if let Some(Ok(line)) = reader.lines().next() {
+            let value: i32 = match line.trim().parse() {
+                Ok(v) => v,
+                Err(_) => panic!("Failed to parse {}", file_path),
+            };
+
+            value
+        } else {
+            panic!("empty {}", file_path);
+        }
+    }
+
+    // Write an i32 to a file in procfs.
+    fn write_procfs(&self, file_path: &str, value: i32) {
+        let mut file = File::create(file_path).expect(&format!("Failed to open {}", file_path));
+        file.write_all(value.to_string().as_bytes())
+            .expect(&format!("Failed to write to {}", file_path));
+    }
+
+    // Save all the sysctl VM settings in the internal state.
+    fn save(&self) {
+        let compact_unevictable_allowed = "/proc/sys/vm/compact_unevictable_allowed";
+        let value = self.read_procfs(compact_unevictable_allowed);
+        unsafe {
+            VM.compact_unevictable_allowed = value;
+        };
+        self.write_procfs(compact_unevictable_allowed, 0);
```

Comment:
I wonder whether BPF part can detect usersched thread getting blocked on a page fault (or anything which isn't kosher) and then just direct dispatch everything till the usersched gets unblocked. This would require BPF part to keep track of what's queued on the usersched side. Hmm... that's already tracked per-rq in sched_ext core. Maybe we just need to something similar to `scx_bpf_reenq_local()` which re-enqueues everything which is not queued on one of the dsq's.

Reply:
> I wonder whether BPF part can detect usersched thread getting blocked on a page fault (or anything which isn't kosher) and then just direct dispatch everything till the usersched gets unblocked. This would require BPF part to keep track of what's queued on the usersched side. Hmm... that's already tracked per-rq in sched_ext core. Maybe we just need to something similar to `scx_bpf_reenq_local()` which re-enqueues everything which is not queued on one of the dsq's.

Do you think userfaultfd can help with that? What if the usersched creates a separate userfaultfd thread, when a page fault happens it will be intercepted by the userfaultfd thread that can set a flag in the BPF part, that would temporarily switch it into a "reschedule everything without usersched" mode and then switch back to "regular" mode once the resched is done.

Reply:
I have never used userfaultfd and am not sure, but from glancing the manpage, it *looks* liked it'd need a thread to be runnable and if the scheduler thread is already blocked on fault, I guess that thread would need a special treatment too to deliver the notification to the BPF.

Hmm... if there's a tracepoint or fentry probe we can attach to monitor the relevant page faults, we can filter it by the PID and notify the BPF scheduler right away skipping going through userspace thread for userfaultfd. See `tp_task_rename()` and `scheduler_tick_fentry()` in `scx_layered` for examples. Would this be an option?

Reply:
Right, userfault thread needs to run, but we could save the pid and dispatch it immediately in the BPF part from `.enqueue`, allowing to always bypass the user-space scheduler.

But maybe the tracepoint/fentry way is easier, hm... I'll take a look.